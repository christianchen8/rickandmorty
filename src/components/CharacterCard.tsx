import Image from "next/image";
import { useShallow } from "zustand/react/shallow";
import { useCharacterStore } from "@/stores/characters";

interface CharacterCardTypes {
  character: Character;
  setSelectedCharacter: (character: Character | null) => void;
}

export function CharacterCard({
  character,
  setSelectedCharacter,
}: CharacterCardTypes) {
  const { status, name, species, image } = character;

  const [selectedFirstCharacter, selectedSecondCharacter] = useCharacterStore(
    useShallow((state) => [
      state.selectedFirstCharacter,
      state.selectedSecondCharacter,
    ])
  );

  const getColor = () => {
    if (status === "Dead") return "bg-red-500";
    if (status === "Alive") return "bg-green-500";
    return "bg-gray-500";
  };

  return (
    <div
      data-testid="container-card"
      className={`flex w-full border rounded-xl hover:shadow-lg cursor-pointer 
      ${selectedFirstCharacter === character ? "bg-blue-100" : ""}
      ${selectedSecondCharacter === character ? "bg-yellow-100" : ""}
      `}
      onClick={() => {
        selectedSecondCharacter === character ||
        selectedFirstCharacter === character
          ? setSelectedCharacter(null)
          : setSelectedCharacter(character);
      }}
    >
      <div className="h-full min-w-[40%] relative rounded-l-xl">
        <Image
          fill
          src={image}
          data-testid="character-image"
          alt="character-image"
          className="object-cover  rounded-l-xl"
        />
      </div>
      <div className="flex-col justify-center flex p-4 min-w-[60%] text-black ">
        <h1 className="leading-4 font-semibold">{name}</h1>

        <div className="mt-2 w-full flex items-center justify-start">
          <div
            className={`h-[16px] w-[16px] rounded-full mr-2 ${getColor()}`}
          ></div>
          <p>
            {status} - {species}
          </p>
        </div>
      </div>
    </div>
  );
}
