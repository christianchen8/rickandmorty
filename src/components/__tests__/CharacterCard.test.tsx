import "@testing-library/jest-dom";
import { render, screen, fireEvent } from "@testing-library/react";
import { CharacterCard } from "../CharacterCard";
import { character } from "@/mocks/character";

describe("Character Card", () => {

  it("should render correctly", () => {
    const { container } = render(
      <CharacterCard character={character} setSelectedCharacter={() => {}} />
    );
    expect(container).toMatchSnapshot();
  });

  it("should render Image correctly", () => {
    render(
      <CharacterCard character={character} setSelectedCharacter={() => {}} />
    );
    const image = screen.getByTestId("character-image");
    expect(image).toHaveAttribute("alt", "character-image");
  });

  test("should call closeRightSection callback", () => {
    const setSelectedCharacter = jest.fn();
    const { getByTestId } = render(
      <CharacterCard character={character} setSelectedCharacter={setSelectedCharacter} />
    );
    fireEvent.click(getByTestId("container-card"));
    expect(setSelectedCharacter).toHaveBeenCalled();
  });
});
