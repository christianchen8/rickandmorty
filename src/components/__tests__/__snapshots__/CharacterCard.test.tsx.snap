// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character Card should render correctly 1`] = `
<div>
  <div
    class="flex w-full border rounded-xl hover:shadow-lg cursor-pointer 
      
      
      "
    data-testid="container-card"
  >
    <div
      class="h-full min-w-[40%] relative rounded-l-xl"
    >
      <img
        alt="character-image"
        class="object-cover  rounded-l-xl"
        data-nimg="fill"
        data-testid="character-image"
        decoding="async"
        loading="lazy"
        sizes="100vw"
        src="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=3840&q=75"
        srcset="/_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Frickandmortyapi.com%2Fapi%2Fcharacter%2Favatar%2F1.jpeg&w=3840&q=75 3840w"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
    </div>
    <div
      class="flex-col justify-center flex p-4 min-w-[60%] text-black "
    >
      <h1
        class="leading-4 font-semibold"
      >
        Rick Sanchez
      </h1>
      <div
        class="mt-2 w-full flex items-center justify-start"
      >
        <div
          class="h-[16px] w-[16px] rounded-full mr-2 bg-green-500"
        />
        <p>
          Alive
           - 
          Human
        </p>
      </div>
    </div>
  </div>
</div>
`;
